libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -3979,6 +3979,29 @@ const WaveClip* WaveTrack::GetClipAtTime(double time) const
    return p != clips.rend() ? *p : nullptr;
 }
 
+auto WaveTrack::GetWideClipAtTime(double time) -> IntervalHolder
+{
+   const auto clips = SortedIntervalArray();
+   auto p = std::find_if(
+      clips.rbegin(), clips.rend(), [&](const auto &pClip) {
+         return pClip->WithinPlayRegion(time);
+      });
+
+   // When two clips are immediately next to each other, the GetPlayEndTime() of the first clip
+   // and the GetPlayStartTime() of the second clip may not be exactly equal due to rounding errors.
+   // If "time" is the end time of the first of two such clips, and the end time is slightly
+   // less than the start time of the second clip, then the first rather than the
+   // second clip is found by the above code. So correct this.
+   if (p != clips.rend() && p != clips.rbegin() &&
+      time == (*p)->GetPlayEndTime() &&
+      (*p)->GetClip(0)->SharesBoundaryWithNextClip((*(p-1))->GetClip(0).get()))
+   {
+      --p;
```

Comment:
Can this ever be executed?
Say `clipB->WithinPlayRegion(t)` returns true, it then becomes `p`. Then further down you test that `t == p->GetPlayEndTime()`, which can never be true, since `p->WithinPlayRegion(t) == true` implies that `t < p->GetPlayEndTime()` (or there is a bug).

Reply:
I want to make only a conservative refactoring that changes the iteration to interval major.

I did not spend my time reasoning through whether the old extra logic in `GetClipAtTime` is still important.  I assumed it was.

If you think this should simplify, and can prove it, and can also understand why @DavidBailes put that extra logic there, and can convince him too that it is no longer needed -- then I suggest you make a PR for that purpose.

But I believe that should not block approval of this more urgent cleanup.

Reply:
> Can this ever be executed? Say `clipB->WithinPlayRegion(t)` returns true, it then becomes `p`. Then further down you test that `t == p->GetPlayEndTime()`, which can never be true, since `p->WithinPlayRegion(t) == true` implies that `t < p->GetPlayEndTime()` (or there is a bug).

Did you notice that the iteration through the clips was in backward order?

Reply:
https://github.com/audacity/audacity/pull/5586#discussion_r1404086453 intended to prove that already, and to my sense did. From what I read you haven't examined if that extra old logic was useful or not, and yet you introduced a new method replicating this. If `GetIntervalAtTime` didn't exist before I wouldn't mind so much, but it does, and introducing an alternative and even replacing existing usages suggests there's something wrong with it.

Reply:
> Did you notice that the iteration through the clips was in backward order?

I don't see how this impacts the reasoning in https://github.com/audacity/audacity/pull/5586#discussion_r1404086453.

Reply:
> Can this ever be executed? Say `clipB->WithinPlayRegion(t)` returns true, it then becomes `p`. Then further down you test that `t == p->GetPlayEndTime()`, which can never be true, since `p->WithinPlayRegion(t) == true` implies that `t < p->GetPlayEndTime()` (or there is a bug).

@saintmatthieu 
The last time I changed this this function the initial lines were:
```
const auto clips = SortedClipArray();
   auto p = std::find_if(clips.rbegin(), clips.rend(), [&] (WaveClip* const& clip) {
      return time >= clip->GetStartTime() && time <= clip->GetEndTime(); });
```
Since then, it has been changed so that if a time is equal to the time of the end of the clip it is not considered to be in the clip.
So as you say, the lines that correct for rounding errors are no longer necessary.

Reply:
> > Can this ever be executed? Say `clipB->WithinPlayRegion(t)` returns true, it then becomes `p`. Then further down you test that `t == p->GetPlayEndTime()`, which can never be true, since `p->WithinPlayRegion(t) == true` implies that `t < p->GetPlayEndTime()` (or there is a bug).
> 
> @saintmatthieu The last time I changed this this function the initial lines were:
> 
> ```
> const auto clips = SortedClipArray();
>    auto p = std::find_if(clips.rbegin(), clips.rend(), [&] (WaveClip* const& clip) {
>       return time >= clip->GetStartTime() && time <= clip->GetEndTime(); });
> ```
> 
> Since then, it has been changed so that if a time is equal to the time of the end of the clip it is not considered to be in the clip. So as you say, the lines that correct for rounding errors are no longer necessary. (And the times of the start and end of clips are now rounded to multiples of samples.)

You made changes several years ago, and I don't remember the motivating use case to test.  Do you?  Was it clip movement with the keys?

Reply:
Yes, it was clip movement using keystrokes. It was to ensure if a user moved to the end of a clip which was immediately adjacent to a following clip, that the keystrokes to move clips would move the second clip and not the first. The definition of  WaveClip::WithinPlayRegion() ensures that this will happen.